### ISSUE TITLE
Investigate nosetests SystemExit error under Python 3.4+

### ISSUE THREAD

**Issue by pzrq** [mathiasertl/fabric#6]
Followed instructions at:
http://docs.fabfile.org/en/1.10/running_tests.html

I get `SystemExit` causing `FAILED` for `test_should_use_sentinel_for_tasks_that_errored` with pyenv-installed Python 3.4.3 and homebrew-installed Python 3.5.0. I do not get the same failure under the system Python 2.7.10. Each of these is used in a new virtualenv under OSX El Capitan with all updates installed.

Example test runner output (pyenv-installed Python 3.4.3):

```
(fabric343)pzrq@Peters-Mac-mini:~/Projects/fabric$ nosetests tests/
/Users/pzrq/.virtualenvs/fabric343/lib/python3.4/site-packages/nose/plugins/manager.py:395: RuntimeWarning: Unable to load plugin color = rudolf:ColorOutputPlugin: invalid syntax (rudolf.py, line 787)
  RuntimeWarning)
............................................................................................................................................................................................[username@127.0.0.1:2200] err: stderr.[username@127.0.0.1:2200] err: stderr.........................................................Traceback (most recent call last):
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 561, in process_request_thread
    self.finish_request(request, client_address)
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 320, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 631, in __init__
    self.handle()
  File "/Users/pzrq/Projects/fabric/tests/server.py", line 366, in handle
    if self.sudo_prompt and not self.sudo_password():
  File "/Users/pzrq/Projects/fabric/tests/server.py", line 435, in sudo_password
    self.channel.send(env.sudo_prompt)
  File "/Users/pzrq/.virtualenvs/fabric343/src/paramiko/paramiko/channel.py", line 710, in send
    return self._send(s, m)
  File "/Users/pzrq/.virtualenvs/fabric343/src/paramiko/paramiko/channel.py", line 1070, in _send
    raise socket.error('Socket is closed')
OSError: Socket is closed
..Traceback (most recent call last):
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 561, in process_request_thread
    self.finish_request(request, client_address)
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 320, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/Users/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 631, in __init__
    self.handle()
  File "/Users/pzrq/Projects/fabric/tests/server.py", line 366, in handle
    if self.sudo_prompt and not self.sudo_password():
  File "/Users/pzrq/Projects/fabric/tests/server.py", line 435, in sudo_password
    self.channel.send(env.sudo_prompt)
  File "/Users/pzrq/.virtualenvs/fabric343/src/paramiko/paramiko/channel.py", line 710, in send
    return self._send(s, m)
  File "/Users/pzrq/.virtualenvs/fabric343/src/paramiko/paramiko/channel.py", line 1070, in _send
    raise socket.error('Socket is closed')
OSError: Socket is closed
............................................................
Warning: 'thisisnotavalidtaskname' is not callable or a valid task name

..........E..........................................................
Fatal error: Test

Aborting.
.....................
======================================================================
ERROR: Tasks which errored but didn't abort should contain an eg NetworkError
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/pzrq/.virtualenvs/fabric343/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/pzrq/Projects/fabric/tests/test_tasks.py", line 391, in test_should_use_sentinel_for_tasks_that_errored
    retval = execute(task, hosts=[host_string])
  File "/Users/pzrq/Projects/fabric/fabric/tasks.py", line 388, in execute
    multiprocessing
  File "/Users/pzrq/Projects/fabric/fabric/tasks.py", line 278, in _execute
    return task.run(*args, **kwargs)
  File "/Users/pzrq/Projects/fabric/fabric/tasks.py", line 175, in run
    return self.wrapped(*args, **kwargs)
  File "/Users/pzrq/Projects/fabric/tests/test_tasks.py", line 388, in task
    run("whoops")
  File "/Users/pzrq/Projects/fabric/fabric/network.py", line 656, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/Users/pzrq/Projects/fabric/fabric/operations.py", line 1070, in run
    shell_escape=shell_escape)
  File "/Users/pzrq/Projects/fabric/fabric/operations.py", line 931, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/Users/pzrq/Projects/fabric/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/Users/pzrq/Projects/fabric/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/Users/pzrq/Projects/fabric/fabric/network.py", line 158, in __getitem__
    self.connect(key)
  File "/Users/pzrq/Projects/fabric/fabric/network.py", line 150, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/Users/pzrq/Projects/fabric/fabric/network.py", line 539, in connect
    sys.exit(0)
nose.proxy.SystemExit: 0
-------------------- >> begin captured stdout << ---------------------


--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 398 tests in 108.120s

FAILED (errors=1)

```

I should be able to investigate Ubuntu 14.04.1 on my laptop later tonight to see if it's more likely to be Python 3 or OSX related. 

Thought it would be a good idea to try to clean up master before looking into a test and fix for #5

**Comment by pzrq** [comment]
Reproduced under Ubuntu 14.04 so it's most likely a Python 3.4+ issue:

```
(fabric3)pzrq@pzrq-ubox:~/Projects/fabric$ nosetests tests/
/home/pzrq/.virtualenvs/fabric3/lib/python3.4/site-packages/nose/plugins/manager.py:395: RuntimeWarning: Unable to load plugin color = rudolf:ColorOutputPlugin: invalid syntax (rudolf.py, line 787)
  RuntimeWarning)
............................................................................................................................................................................................[username@127.0.0.1:2200] err: stderr.[username@127.0.0.1:2200] err: stderr.........................................................Traceback (most recent call last):
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 561, in process_request_thread
    self.finish_request(request, client_address)
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 320, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 631, in __init__
    self.handle()
  File "/home/pzrq/Projects/fabric/tests/server.py", line 366, in handle
    if self.sudo_prompt and not self.sudo_password():
  File "/home/pzrq/Projects/fabric/tests/server.py", line 435, in sudo_password
    self.channel.send(env.sudo_prompt)
  File "/home/pzrq/.virtualenvs/fabric3/src/paramiko/paramiko/channel.py", line 710, in send
    return self._send(s, m)
  File "/home/pzrq/.virtualenvs/fabric3/src/paramiko/paramiko/channel.py", line 1070, in _send
    raise socket.error('Socket is closed')
OSError: Socket is closed
..Traceback (most recent call last):
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 561, in process_request_thread
    self.finish_request(request, client_address)
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 320, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/pzrq/Projects/fabric/tests/Python26SocketServer.py", line 631, in __init__
    self.handle()
  File "/home/pzrq/Projects/fabric/tests/server.py", line 366, in handle
    if self.sudo_prompt and not self.sudo_password():
  File "/home/pzrq/Projects/fabric/tests/server.py", line 435, in sudo_password
    self.channel.send(env.sudo_prompt)
  File "/home/pzrq/.virtualenvs/fabric3/src/paramiko/paramiko/channel.py", line 710, in send
    return self._send(s, m)
  File "/home/pzrq/.virtualenvs/fabric3/src/paramiko/paramiko/channel.py", line 1070, in _send
    raise socket.error('Socket is closed')
OSError: Socket is closed
............................................................
Warning: 'thisisnotavalidtaskname' is not callable or a valid task name

..........E..........................................................
Fatal error: Test

Aborting.
.....................
======================================================================
ERROR: Tasks which errored but didn't abort should contain an eg NetworkError
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pzrq/.virtualenvs/fabric3/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/pzrq/Projects/fabric/tests/test_tasks.py", line 391, in test_should_use_sentinel_for_tasks_that_errored
    retval = execute(task, hosts=[host_string])
  File "/home/pzrq/Projects/fabric/fabric/tasks.py", line 388, in execute
    multiprocessing
  File "/home/pzrq/Projects/fabric/fabric/tasks.py", line 278, in _execute
    return task.run(*args, **kwargs)
  File "/home/pzrq/Projects/fabric/fabric/tasks.py", line 175, in run
    return self.wrapped(*args, **kwargs)
  File "/home/pzrq/Projects/fabric/tests/test_tasks.py", line 388, in task
    run("whoops")
  File "/home/pzrq/Projects/fabric/fabric/network.py", line 656, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/home/pzrq/Projects/fabric/fabric/operations.py", line 1070, in run
    shell_escape=shell_escape)
  File "/home/pzrq/Projects/fabric/fabric/operations.py", line 931, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/home/pzrq/Projects/fabric/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/home/pzrq/Projects/fabric/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/home/pzrq/Projects/fabric/fabric/network.py", line 158, in __getitem__
    self.connect(key)
  File "/home/pzrq/Projects/fabric/fabric/network.py", line 150, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/home/pzrq/Projects/fabric/fabric/network.py", line 539, in connect
    sys.exit(0)
nose.proxy.SystemExit: 0
-------------------- >> begin captured stdout << ---------------------


--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 398 tests in 107.411s

FAILED (errors=1)
```

**Comment by pzrq** [comment]
This is detecting a bug in Paramiko which should be resolved by: https://github.com/paramiko/paramiko/pull/616

**Comment by mathiasertl** [comment]
What exactly does this mean: do you think this bug will be fixed with a new paramiko version?

**Comment by pzrq** [comment]
@mathiasertl Yes if https://github.com/paramiko/paramiko/pull/616 (or something similar) was merged to Paramiko's master and an updated Paramiko version included in Fabric3's requirements, then it should make this Fabric3 unit test pass under Python 3 (at least in my testing it passed when I hacked that fix into Paramiko in one of my virtualenvs). 

Hence the `test_should_use_sentinel_for_tasks_that_errored` error doesn't feel like a bug with Fabric3. 

In the mean time something like `@unittest.skipIf(PY3)` could be added though having the issue open for others to discuss solutions is probably more important at this time.
